app/controllers/actions/AuthAction.scala

```diff
@@ -0,0 +1,110 @@
+
+      override val parser = bodyParser
+      override val executionContext = ec
+
+      override def invokeBlock[A](request: Request[A], block: UserRequest[A] => Future[Result]): Future[Result] = {
+
+        implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequest(request)
+
+        implicit val req = request
+
+        authorised().retrieve(Retrievals.affinityGroup and Retrievals.allEnrolments and Retrievals.internalId and Retrievals.credentials) {
+
+          case Some(Organisation) ~ enrolments ~ Some(internalId) ~ Some(credentials) =>
+            logger.debug("[invokeBlock] Checking for EMCS Enrolment")
+            checkOrganisationEMCSEnrolment(ern, enrolments, internalId, credentials.providerId)(block)
+
+          case Some(Organisation) ~ _ ~ None ~ _ =>
+            logger.warn("[invokeBlock] InternalId could not be retrieved from Auth")
+            Future.successful(Unauthorized)
+
+          case Some(Organisation) ~ _ ~ _ ~ None =>
+            logger.warn("[invokeBlock] Credentials could not be retrieved from Auth")
+            Future.successful(Unauthorized)
+
+          case Some(affinityGroup) ~ _ ~ _ ~ _ =>
+            logger.warn(s"[invokeBlock] User has incompatible AffinityGroup of '$affinityGroup'")
+            Future.successful(Unauthorized)
+
+          case _ =>
+            logger.warn(s"[invokeBlock] User has no AffinityGroup")
+            Future.successful(Unauthorized)
+
+        } recover {
+          case x: NoActiveSession =>
+            logger.debug(s"[invokeBlock] NoActiveSession Exception with reason: ${x.reason}")
+            Unauthorized
+          case x: AuthorisationException =>
+            logger.debug(s"[invokeBlock] Authorisation Exception ${x.reason}")
+            Forbidden
+        }
+      }
+    }
+
+  private def checkOrganisationEMCSEnrolment[A](ernFromUrl: String,
+                                                enrolments: Enrolments,
+                                                internalId: String,
+                                                credId: String
+                                               )(block: UserRequest[A] => Future[Result])
+                                               (implicit request: Request[A]): Future[Result] =
+    enrolments.enrolments.filter(enrolment => enrolment.key == EnrolmentKeys.EMCS_ENROLMENT) match {
+      case emcsEnrolments if emcsEnrolments.isEmpty =>
+        logger.debug(s"[checkOrganisationEMCSEnrolment] No ${EnrolmentKeys.EMCS_ENROLMENT} enrolment found")
+        Future.successful(Forbidden)
+      case emcsEnrolments =>
+        emcsEnrolments.find(_.identifiers.exists(ident => ident.key == EnrolmentKeys.ERN && ident.value == ernFromUrl)) match {
+          case Some(enrolment) if enrolment.isActivated =>
+            block(UserRequest(request, ernFromUrl, internalId, credId, emcsEnrolments.flatMap(_.identifiers.map(_.value))))
```

Comment:
I'd probably add `allUserERNs =` to the final field in this UserRequest, as it's a bit clearer. Happy for you to leave it though